### PR TITLE
Gdax

### PR DESCRIPTION
**Pull request by jaredbroad**

### CODE REVIEW

File: Brokerages/GDAX/GDAXBrokerage.cs [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+                    CachedOrderIDs.TryRemove(order.Id, out outOrder);
+                }
+
+                Log.Trace("GDAXBrokerage.PlaceOrder: Order completed successfully orderid:" + order.Id.ToString());
+                return true;
+
+            }
+
+            OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, 0, "GDAX Order Event") { Status = OrderStatus.Invalid });
+            Log.Trace("GDAXBrokerage.PlaceOrder: Order failed Order Id: " + order.Id + " timestamp:" + order.Time + " quantity: " + order.Quantity.ToString()
```
Comment (jaredbroad) [inline Brokerages/GDAX/GDAXBrokerage.cs]: Log.Error for failed error messages; & probably returning some information to the user-side via brokerage messaging handler.

File: Brokerages/GDAX/GDAXBrokerage.cs [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        }
+
+        /// <summary>
+        /// Cancels an order
+        /// </summary>
+        /// <param name="order"></param>
+        /// <returns></returns>
+        public override bool CancelOrder(Orders.Order order)
+        {
+            List<bool> success = new List<bool>();
```
Comment (jaredbroad) [inline Brokerages/GDAX/GDAXBrokerage.cs]: Using var where possible

File: Brokerages/GDAX/GDAXBrokerage.cs [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+            var req = new RestRequest("/orders?status=active", Method.GET);
+            GetAuthenticationToken(req);
+            var response = RestClient.Execute(req);
+            if (response != null)
+            {
+                foreach (var item in JsonConvert.DeserializeObject<Messages.Order[]>(response.Content))
+                {
+
+                    decimal conversionRate;
+                    if (!item.ProductId.EndsWith("USD", StringComparison.InvariantCultureIgnoreCase))
```
Comment (jaredbroad) [inline Brokerages/GDAX/GDAXBrokerage.cs]: nice

File: Brokerages/GDAX/GDAXBrokerage.cs [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+                    {
+
+                        var baseSymbol = (item.ProductId.Substring(0, 3) + "USD").ToLower();
```
Comment (jaredbroad) [inline Brokerages/GDAX/GDAXBrokerage.cs]: We'll soon need to figure out how to solve the crypto-symbol issue =) Lets kick that one down the road a little....

File: Brokerages/GDAX/GDAXBrokerage.cs [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+
+            foreach (var item in JsonConvert.DeserializeObject<Messages.Account[]>(response.Content))
+            {
+                if (item.Balance > 0)
+                {
+                    if (item.Currency == "USD")
+                    {
+                        list.Add(new Securities.Cash(item.Currency, item.Balance, 1));
+                    }
+                    else if (new[] {"GBP", "EUR" }.Contains(item.Currency))
```
Comment (jaredbroad) [inline Brokerages/GDAX/GDAXBrokerage.cs]: elegant,

File: Brokerages/GDAX/GDAXBrokerage.Messaging.cs [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+                    Thread.Sleep(delay);
+                    if (token.IsCancellationRequested) break;
+                }
+                Log.Trace("PollLatestTick: " + "Stopped polling for ticks: " + symbol.Value.ToString());
+            }, token, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+        }
+
+        private decimal GetConversionRate(string currency)
+        {
+            var response = RateClient.Execute(new RestSharp.RestRequest(Method.GET));
```
Comment (jaredbroad) [inline Brokerages/GDAX/GDAXBrokerage.Messaging.cs]: this is a no-op right? Assume currently only converts USD balances?

File: Brokerages/GDAX/GDAXBrokerage.cs [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+            {
+                var copy = Ticks.ToArray();
+                Ticks.Clear();
+                return copy;
+            }
+        }
+        #endregion
+
+        /// <summary>
+        /// Reteives the fee for a given order
```
Comment (jaredbroad) [inline Brokerages/GDAX/GDAXBrokerage.cs]: Minor, spelling

File: Common/Brokerages/BrokerageName.cs [hunk 8]
```diff
@@ -54,6 +54,12 @@ public enum BrokerageName
         /// <summary>
         /// Transaction and submit/execution rules will use bitfinex models
         /// </summary>
-        Bitfinex
+        Bitfinex,
+
+        /// <summary>
+        /// Transaction and submit/execution rules will use gdax models
+        /// </summary>
+        GDAX = 12
```
Comment (jaredbroad) [inline Common/Brokerages/BrokerageName.cs]: Curious why 12 here?

File: Common/Brokerages/GDAXBrokerageModel.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        /// <summary>
+        /// GDAX global leverage rule
+        /// </summary>
+        /// <param name="security"></param>
+        /// <returns></returns>
+        public override decimal GetLeverage(Security security)
+        {
+            return 3m;
```
Comment (jaredbroad) [inline Common/Brokerages/GDAXBrokerageModel.cs]: Perhaps 1 to start? By memory you mentioned in chat the code doesn't support leverage yet?

File: Common/Market.cs [hunk 10]
```diff
@@ -43,7 +43,8 @@ public static class Market
             Tuple.Create(CBOT, 8),
             Tuple.Create(ICE, 9),
             Tuple.Create(CBOE, 10),
-            Tuple.Create(NSE,11)
+            Tuple.Create(NSE, 11),
+            Tuple.Create(GDAX, 12)
```
Comment (jaredbroad) [inline Common/Market.cs]: This is why its 12 👍 Nice idea

File: Data/market-hours/market-hours-database.json [hunk 11]
```diff
@@ -10082,7 +10082,7 @@
           "state": "market"
         }
       ],
-      "saturday": [],
+      "saturday": [ ],
```
Comment (jaredbroad) [inline Data/market-hours/market-hours-database.json]: Lots of white space changes here

File: Tests/Brokerages/GDAX/GDAXBrokerageTests.cs [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+        Mock<IAlgorithm> _algo = new Mock<IAlgorithm>();
+        string _orderData;
+        string _orderByIdData;
+        string _openOrderData;
+        string _matchData;
+        string _accountsData;
+        string _holdingData;
+        string _tickerData;
+        Symbol _symbol;
+        const string _brokerId = "d0c5340b-6d6c-49d9-b567-48c4bfca13d2";
```
Comment (jaredbroad) [inline Tests/Brokerages/GDAX/GDAXBrokerageTests.cs]: Interesting, what are these?

File: Tests/QuantConnect.Tests.csproj [hunk 13]
```diff
@@ -12,7 +12,6 @@
     <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\</SolutionDir>
-    <RestorePackages>true</RestorePackages>
```
Comment (jaredbroad) [inline Tests/QuantConnect.Tests.csproj]: Guess this should remain in the csproj?

File: Common/Brokerages/GDAXBrokerageModel.cs (hunk 9 above)
Comment (jameschch) [reply Common/Brokerages/GDAXBrokerageModel.cs]: I've not been able to test leverage. In theory nothing remains to be done to support leverage, but it sounds like we will need to disable for now.

File: Tests/Brokerages/GDAX/GDAXBrokerageTests.cs (hunk 12 above)
Comment (jameschch) [reply Tests/Brokerages/GDAX/GDAXBrokerageTests.cs]: Text files supply sample json to unit test the websockets pipeline abstracted and offline (and part of CI). Pretty much my standard practice. In this case, code coverage is fairly decent.

File: Tests/QuantConnect.Tests.csproj (hunk 13 above)
Comment (jameschch) [reply Tests/QuantConnect.Tests.csproj]: This appeared to be the only csproj that had it and caused problems in vs2015.

File: Brokerages/GDAX/GDAXBrokerage.Messaging.cs (hunk 6 above)
Comment (jameschch) [reply Brokerages/GDAX/GDAXBrokerage.Messaging.cs]: Yes, this workarounds the conversion rate not found error for non USD denominated currencies. The alternative was to require an FXCM or Oanda subscription or use the compound rate from two tickers. This issue might deserve a resolution that goes beyond one brokerage.

**Pull request merged**